In [168]:
import pandas as pd
import numpy as np
from pprint import pprint
from sklearn.metrics import accuracy_score

# Import dataset from the csv file
df = pd.read_csv('diabetes.csv')
X = df.iloc[:, :-1].values
y = df.iloc[:, -1].values

from sklearn.model_selection import train_test_split

#Splitting the data into test and train data subsets

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.4, random_state=54656323, shuffle=True)

#Perceptron class

class Perceptron:
    
    def __init__(self, learning_rate = 0.1, n_iter = 2000):
        
        self.learning_rate = learning_rate
        self.n_iter = n_iter
        self.weights = None
        self.bias = None

    #initialises weights, does dot product with subsets of X
    def fit(self, X, y ):
        n_rows, n_cols = X.shape
        self.weights = np.zeros(n_cols)
        self.bias = 0
        
        for j in range(1000):
            
            for i in range(n_rows):
                y_proj = np.dot(X[i], self.weights) + self.bias
                if y_proj >= 0:
                    activation = 1
                else:
                    activation = 0

                update = self.learning_rate * (y[i]-activation)
                self.weights += update*X[i]
                self.bias += update

    def predict(self, X):
        
        y_pred = np.dot(X, self.weights)+self.bias
        return (y_pred >= 0).astype(int)
        
perceptron = Perceptron(learning_rate=0.8, n_iter=2000)
perceptron.fit(X_train, y_train)
y_pred_train = perceptron.predict(X_train)



y_pred_test = perceptron.predict(X_test)
accuracy = accuracy_score(y_test, y_pred_test)
print("Accuracy:", accuracy)


        
    
    


Accuracy: 0.6334056399132321


learning rate, test_train_split is constant (0.001, 0.4)
n_iterations  = 20 | accuracy = 0.36
n_iterations = 200 | accuracy = 0.65
n_iterations = 2000 | accuracy = 0.63
n_iterations = 20000 | accuracy = 0.0.64

learning rate, test_train_split is constant (0.1, 0.4)
n_iterations  = 20 | accuracy = 0.39
n_iterations = 200 | accuracy = 0.64
n_iterations = 2000 | accuracy = 0.41
n_iterations = 20000 | accuracy = 0.64


n_iterations, test_train_split is constant (2000, 0.4)
learning_rate = 0.0001 |  accuracy = 0.66
learning_rate = 0.001 |  accuracy = 0.67
learning_rate = 0.01 | accuracy = 0.65
learning_rate = 0.1 | accuracy = 0.58
learning_rate = 0.2 | accuracy = 0.64
learning_rate = 0.6 | accuracy = 0.66
learning_rate = 0.8 | accuracy = 0.63